components/engine_traits/src/import.rs

```diff
@@ -1,15 +1,201 @@
 // Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
 
-use crate::errors::Result;
+use std::{
+    collections::BTreeMap,
+    sync::{Arc, Mutex, RwLock},
+};
+
+use crate::{errors::Result, Range};
+
+/// Currently used to ensure mutual exclusion between
+/// compaction filter GC and ingest SST operations.
+#[derive(Debug)]
+pub struct RangeLatch {
+    /// A BTreeMap storing active range locks.
+    /// Key: The start key of the range (sorted order).
+    /// Value: A tuple of:
+    ///   - `Arc<Mutex<()>>`: The latch object for this range.
+    ///   - `(Vec<u8>, Vec<u8>)`: The actual range definition (start_key,
+    ///     end_key).
+    range_latches: RwLock<BTreeMap<Vec<u8>, (Arc<Mutex<()>>, (Vec<u8>, Vec<u8>))>>,
+
+    #[cfg(any(test, feature = "testexport"))]
+    latch_history: Mutex<Vec<String>>,
+}
+
+impl RangeLatch {
+    pub fn new() -> Self {
+        Self {
+            range_latches: RwLock::new(BTreeMap::new()),
+            #[cfg(any(test, feature = "testexport"))]
+            latch_history: Mutex::new(Vec::new()),
+        }
+    }
+
+    /// Acquires a range latch for the specified `[start_key, end_key)`.
+    /// This function ensures that a range latch is only granted when there are
+    /// no overlapping ranges already latched. If overlaps exist, the
+    /// function waits until those latches are released before proceeding.
+    ///
+    ///  The concurrency is very low because only a few compaction filter
+    /// threads (limited by `max_background_jobs`) and a single region
+    /// worker thread run concurrently, with each thread holding at most one
+    /// latch.
+    ///
+    /// Overlaps are determined by iterating through all active latches stored
+    /// in `range_latches` and checking whether the specified range
+    /// conflicts with any of them. While this approach may seem
+    /// inefficient, the overhead of linear iteration is negligible in practice
+    /// due to the low concurrency of the system.
+    ///
+    /// Live-locks can occur in this implementation. For example, a thread
+    /// attempting to acquire a latch for the range `[a, z)` might encounter
+    /// that `[a, b)` and `[b, c)` are already latched by other threads. The
+    /// thread will first wait for `[a, b)` to be released, then proceed to wait
+    /// for `[b, c)`. However, during the wait for `[b, c)` to be released,
+    /// another thread might acquire `[a, b)` again, forcing the original
+    /// thread to wait for `[a, b)` once more. This process could repeat
+    /// indefinitely if threads continuously "jump the queue." Despite this,
+    /// such live-locks are unlikely to persist for long due to the system's low
+    /// concurrency, as conflicting latches are eventually released, allowing
+    /// threads to make progress naturally.
```

Comment:
When doing non-blocking sst ingestion, the sst ingestion and concurrent write batch(e.g. from compaction filter) are mutually exclusive with each other. That means if a write batch under range [a, z)  is running, all sst ingestion within the same range should be block, and vice verse. So both ingest and write should hold a range mutex here.